[Imaging] Improvements to the Order Independent Transparency feature in HdStream

sirpalee opened the pull request:
### Description of Change(s)
The following PR adds new features and fixes issues related to OIT for HdStream.
- Previously viewport resizing was slow because the OIT buffers were initialized with a default buffer source argument, leading to the creation of large CPU side buffers with the same value. This is now removed, and the new code is using the "Resize" function on the buffers. We don't need to set their default value since it's either going to be cleared, or guaranteed to be overwritten by the OIR Render Pass.
- Removed all the buffer clear for OIT, except the counter buffer. We don't need to clear the buffers before the render since the values of every other OIT buffer is going to be overwritten by the OIT render pass.
- Replaced the insert sort function with a select sorting for the oitResolveImage function. While benchmarking the shader, we found that select sorting is roughly the same performance as insert sorting with 8 samples, but it's significantly faster with more samples (about 2-3 times faster with 32-64 samples). These test results were seen on Centos, with a GTX 1060 (410.78) and the "powerplant" scene with all the faces being partially transparent. Alternatively, it's easy to make the sorting method selectable using an environment variable, if there are cases where select sorting is slower than insert sorting.
- The number of OIT samples per pixel is now a render delegate parameter and can be changed on the fly. To maintain consistency with parameter names across all the different bits of the code, I renamed all the parameter uses to oitNumSamples. Please let me know if you prefer the name being oitSamples instead.
- The PR also includes our fix for #852 . Please let me know if you would like to get it removed from this PR.
- We saw issues with OIT in Maya 2019 and 2018 on Centos, glClearNamedBuffer is not available in those contexts, so we added a simple function that falls back to glClearNamedBufferEXT if glClearNamedBuffer is not available.
- Added an early exit to RenderOutput (in hdx/renderPass.glslfx), if the alpha value of the fragment is small.

### Fixes Issues
- #852

jilliene commented:
Filed as internal issue #USD-5388

lumonix commented:
Thank you for this contribution @sirpalee 
This is great.

Since some of your changes overlap with internal OIT work (will start to land in dev soon) I'll take some of your contributions in manually via internal change-lists instead of merging this request.

We'll wait a bit longer on exposing settings (like the numSamples) until we have completed our on-going volume rendering work in Hydra.

sirpalee commented:
@lumonix sounds good to me!

I updated the PR with a few small changes:
- Fixed a debug message if the oit num samples parameter's type doesn't match.
- Missed a minimum value limit for the number of samples in the resolve task.
- Added a new constructor to HdRenderSettingDescriptor, so we can use .emplace_back on the HdRenderSettingDescriptorList. This way adding new render settings to the HdSt list is a bit more convenient, and less error-prone (when there is a mismatch in the preallocated vector, vs number of settings). I also added a default constructor, so existing code in other render delegates work as well.

poljere commented:
Thanks @sirpalee. We are going to close this PR now, we have internally incorporated some of your changes and you will see them in the coming days in dev, for instance: https://github.com/PixarAnimationStudios/USD/commit/42a7ebf5f5fed3654e8416287099f5c1d4a2900c
